Created listener for worker state change

lc-nyovchev opened the pull request:
Related to: https://github.com/awslabs/amazon-kinesis-client/issues/275

As the Worker itself is a Runnable, it'd be good to know about its progress so we can take some actions or even have some custom logging on top of that.

pfifer commented on src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:
```diff
@@ -866,6 +878,22 @@ protected ShardConsumer buildConsumer(ShardInfo shardInfo, IRecordProcessorFacto
+     * A listener for callbacks on changes worker state
+     */
+    @FunctionalInterface
+    public interface WorkerStateChangeListener {
```
Could you move this to its own file?

lc-nyovchev replied:
Of course

pfifer commented on src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:
```diff
@@ -275,7 +276,8 @@ public Worker(
                 config.getRetryGetRecordsInSeconds(),
-                config.getMaxGetRecordsThreadPool());
+                config.getMaxGetRecordsThreadPool(),
+                new NoOpWorkerStateChangeListener());
```
Could you make this a static constant instead of creating one.

lc-nyovchev replied:
Of course

pfifer commented on src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:
```diff
@@ -347,7 +349,7 @@ public Worker(
                 shardSyncIdleTimeMillis, cleanupLeasesUponShardCompletion, checkpoint, leaseCoordinator, execService,
                 metricsFactory, taskBackoffTimeMillis, failoverTimeMillis, skipShardSyncAtWorkerInitializationIfLeasesExist,
-                shardPrioritization, Optional.empty(), Optional.empty());
+                shardPrioritization, Optional.empty(), Optional.empty(), new NoOpWorkerStateChangeListener());
```
Same as above, can you please make this a constant and use it instead.

lc-nyovchev replied:
will do

pfifer commented on src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:
```diff
@@ -1282,6 +1322,10 @@ public Worker build() {
                     .getProxy(config.getStreamName());
             }
 
+            if (workerStateChangeListener == null){
```
Please match the code style and add a space after the \)

lc-nyovchev replied:
sorry :(. Will do :)

pfifer commented on src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:
```diff
@@ -1282,6 +1322,10 @@ public Worker build() {
             }
 
+            if (workerStateChangeListener == null){
+                workerStateChangeListener = new NoOpWorkerStateChangeListener();
```
Use the constant here as well.

lc-nyovchev commented:
hey @pfifer, removed the conflicts, added some unit tests for the builder and for the state change, removed the extra space to conform with the code style.

lc-nyovchev commented on src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java:
```diff
@@ -1500,6 +1504,105 @@ public void testBuilderWhenKinesisProxyIsSet() {
+        when(taskFuture.get()).thenReturn(taskResult);
+        when(taskResult.isShardEndReached()).thenReturn(true);
+
+        Worker worker = new Worker.Builder()
```
I firmly believe unit tests should strictly stick to the public apis to construct the proper objects, so I tried using only the builder to make the test, and not the visible-for-tests-constructor

pfifer commented:
Thanks for the updates.  I'm starting work on my side to merge the changes.

pfifer commented:
Just one more thing: 

Please confirm that we can use, modify, copy, and redistribute this contribution. Thanks.

lc-nyovchev commented:
I hereby confirm that you guys can use, modify, copy and redistribute this contribution, and any further ones I make :)
